File: core/src/main/java/org/mifos/mobilewallet/core/data/fineract/entity/mapper/TransactionMapper.java
package org.mifos.mobilewallet.core.data.fineract.entity.mapper;

import org.mifos.mobilewallet.core.data.fineract.entity.accounts.savings.SavingsWithAssociations;
import org.mifos.mobilewallet.core.data.fineract.entity.accounts.savings.Transactions;
import org.mifos.mobilewallet.core.domain.model.CheckBoxStatus;
import org.mifos.mobilewallet.core.domain.model.Transaction;
import org.mifos.mobilewallet.core.domain.model.TransactionType;
import org.mifos.mobilewallet.core.utils.DateHelper;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import javax.inject.Inject;

/**
 * Created by naman on 15/8/17.
 */

public class TransactionMapper {

    @Inject
    CurrencyMapper currencyMapper;

    @Inject
    public TransactionMapper() {
    }

    public List<Transaction> transformTransactionList(SavingsWithAssociations
            savingsWithAssociations, List<CheckBoxStatus> filterList) {
        List<Transaction> transactionList = new ArrayList<>();
        Set<TransactionType> checkedFilterSet = getCheckedFilterSet(filterList);

        if (savingsWithAssociations != null && savingsWithAssociations.getTransactions() != null
                && savingsWithAssociations.getTransactions().size() != 0) {

            for (Transactions transactions : savingsWithAssociations.getTransactions()) {
                Transaction transaction = transformInvoice(transactions);
                if (checkedFilterSet.contains(transaction.getTransactionType())) {
                    transactionList.add(transaction);
                }
            }
        }
        return transactionList;
    }

    public List<Transaction> transformTransactionList(SavingsWithAssociations
                                                              savingsWithAssociations) {
        List<Transaction> transactionList = new ArrayList<>();

        if (savingsWithAssociations != null && savingsWithAssociations.getTransactions() != null
                && savingsWithAssociations.getTransactions().size() != 0) {

            for (Transactions transactions : savingsWithAssociations.getTransactions()) {
                Transaction transaction = transformInvoice(transactions);
                transactionList.add(transaction);
            }
        }
        return transactionList;
    }

    public Transaction transformInvoice(Transactions transactions) {
        Transaction transaction = new Transaction();

        if (transactions != null) {

            transaction.setTransactionId(transactions.getId().toString());

            if (transactions.getPaymentDetailData() != null) {
                transaction.setReceiptId(transactions.getPaymentDetailData().getReceiptNumber());
            }

            transaction.setAmount(transactions.getAmount());

            if (transactions.getSubmittedOnDate() != null) {
                transaction.setDate(DateHelper.getDateAsString(transactions.getSubmittedOnDate()));
            }

            transaction.setCurrency(currencyMapper.transform(transactions.getCurrency()));
            transaction.setTransactionType(TransactionType.OTHER);

            if (transactions.getTransactionType().getDeposit()) {
                transaction.setTransactionType(TransactionType.CREDIT);
            }

            if (transactions.getTransactionType().getWithdrawal()) {
                transaction.setTransactionType(TransactionType.DEBIT);
            }

            if (transactions.getTransfer() != null) {
                transaction.setTransferId(transactions.getTransfer().getId());
            }
        }
        return transaction;
    }

    public Set<TransactionType> getCheckedFilterSet(List<CheckBoxStatus> filterList) {
        Set<TransactionType> checkedFilterSet = new HashSet<>();
        for (CheckBoxStatus status : filterList) {
            if (status.getChecked()) {
                if (status.getText().equals("Credit")) {
                    checkedFilterSet.add(TransactionType.CREDIT);
                } else if (status.getText().equals("Debit")) {
                    checkedFilterSet.add(TransactionType.DEBIT);
                } else if (status.getText().equals("Other")) {
                    checkedFilterSet.add(TransactionType.OTHER);
                }
            }
        }
        return  checkedFilterSet;
    }
}

File: core/src/main/java/org/mifos/mobilewallet/core/data/fineract/entity/mapper/TransactionMapper.kt
package org.mifos.mobilewallet.core.data.fineract.entity.mapper

import org.mifos.mobilewallet.core.data.fineract.entity.accounts.savings.SavingsWithAssociations
import org.mifos.mobilewallet.core.data.fineract.entity.accounts.savings.Transactions
import org.mifos.mobilewallet.core.domain.model.Transaction
import org.mifos.mobilewallet.core.domain.model.TransactionType
import org.mifos.mobilewallet.core.utils.DateHelper
import java.util.*
import javax.inject.Inject

/**
 * Created by naman on 15/8/17.
 */
class TransactionMapper @Inject constructor() {

    @Inject
    lateinit var currencyMapper: CurrencyMapper

    fun transformTransactionList(savingsWithAssociations: SavingsWithAssociations?)
            : List<Transaction> {
        var transactionList: MutableList<Transaction> = ArrayList()
        if (savingsWithAssociations?.transactions != null
                && savingsWithAssociations.transactions.isNotEmpty()) {
            for (transactions in savingsWithAssociations.transactions) {
                transactionList.add(transformInvoice(transactions))
            }
        }
        return transactionList
    }

    fun transformInvoice(transactions: Transactions?): Transaction {
        val transaction = Transaction()
        if (transactions != null) {
            transaction.transactionId = transactions.id.toString()
            if (transactions.paymentDetailData != null && transactions.paymentDetailData != null) {
                transaction.receiptId = transactions.paymentDetailData!!.receiptNumber
            }
            transaction.amount = transactions.amount!!
            if (transactions.submittedOnDate != null) {
                transaction.date = DateHelper.getDateAsString(transactions.submittedOnDate)
            }
            transaction.currency = currencyMapper!!.transform(transactions.currency!!)
            transaction.transactionType = TransactionType.OTHER
            if (transactions.transactionType!!.deposit!!) {
                transaction.transactionType = TransactionType.CREDIT
            }
            if (transactions.transactionType!!.withdrawal!!) {
                transaction.transactionType = TransactionType.DEBIT
            }
            if (transactions.transfer != null) {
                transaction.transferId = transactions.transfer!!.id
            }
        }
        return transaction
    }
}